ledger-tool/src/bigtable.rs

```diff
@@ -336,6 +338,250 @@ pub async fn transaction_history(
     Ok(())
 }
 
+struct CopyArgs {
+    from_slot: Slot,
+    to_slot: Option<Slot>,
+
+    source_is_emulator: bool,
+    source_instance_name: String,
+    source_app_profile_id: String,
+    source_credential_path: Option<String>,
+    source_endpoint: Option<String>,
+
+    destination_is_emulator: bool,
+    destination_instance_name: String,
+    destination_app_profile_id: String,
+    destination_credential_path: Option<String>,
+    destination_endpoint: Option<String>,
+
+    force: bool,
+}
+
+impl CopyArgs {
+    pub fn process(arg_matches: &ArgMatches) -> Self {
+        CopyArgs {
+            from_slot: value_t!(arg_matches, "starting_slot", Slot).unwrap_or(0),
+            to_slot: value_t!(arg_matches, "ending_slot", Slot).ok(),
+
+            source_is_emulator: arg_matches.is_present("source_is_emulator"),
+            source_instance_name: value_t_or_exit!(arg_matches, "source_instance_name", String),
+            source_app_profile_id: value_t_or_exit!(arg_matches, "source_app_profile_id", String),
+            source_credential_path: value_t!(arg_matches, "source_credential_path", String).ok(),
+            source_endpoint: value_t!(arg_matches, "source_endpoint", String).ok(),
+
+            destination_is_emulator: arg_matches.is_present("destination_is_emulator"),
+            destination_instance_name: value_t_or_exit!(
+                arg_matches,
+                "destination_instance_name",
+                String
+            ),
+            destination_app_profile_id: value_t_or_exit!(
+                arg_matches,
+                "destination_app_profile_id",
+                String
+            ),
+            destination_credential_path: value_t!(
+                arg_matches,
+                "destination_credential_path",
+                String
+            )
+            .ok(),
+            destination_endpoint: value_t!(arg_matches, "destination_endpoint", String).ok(),
+
+            force: arg_matches.is_present("force"),
+        }
+    }
+}
+
+async fn copy(args: CopyArgs) -> Result<(), Box<dyn std::error::Error>> {
+    let from_slot = args.from_slot;
+    let to_slot = args.to_slot.unwrap_or(from_slot + 1);
+    debug!("from_slot: {}, to_slot: {}", from_slot, to_slot);
+
+    let source_bigtable = {
```

Comment:
consider refactoring the BT instance details to a container type, then this block as a function.  with those changes, you can reuse the same logic for source and destination

Reply:
Do you mean I create a new constructor for BT instance. use emulated_source and source_credential_path as its args and move this logic into the constructor? 🤔

Reply:
[These fields](https://github.com/solana-labs/solana/blob/7ae779120f7af5996526c3a41a0f7bf34fae3b03/ledger-tool/src/bigtable.rs#L345-L349) are common between the source and dest. I'm suggesting to factor them out to a common struct (an instance each for source and dest), the implementing this block as a method on it to remove the replicode between this block and the next